### PR TITLE
Warnings target switch intrinsic asm

### PR DESCRIPTION
**Pull request by venkataram-nv**
Handles #4720 

Also adds a fix to skip generating IR for empty target switches. Previously this would result in generating an unreachable block, which would have necessitated special handling for target switches.

### CODE REVIEW

File: source/slang/slang-ir-reachability.cpp [hunk 1]
```diff
@@ -54,6 +55,45 @@ namespace Slang
...
+                blocks.add(tswitch->getCaseBlock(i));
+
+            IRInst* tmp = inst1->getParent();
+            while (tmp && tmp != tswitch->getParent())
+            {
+                if (blocks.contains(tmp))
+                    return true;
+
+                // Branching instructions
+                if (tmp->firstUse)
```
Comment (csyonghe) [inline source/slang/slang-ir-reachability.cpp:88]: I don't understand what this is tracking.

File: source/slang/slang-ir-reachability.cpp [hunk 2]
```diff
@@ -54,6 +55,45 @@ namespace Slang
...
+        // Special cases
+
+        // Target switches; treat as reachable from its cases
+        if (auto tswitch = as<IRTargetSwitch>(inst2))
+        {
+            HashSet<IRInst*> blocks;
+            for (Slang::UInt i = 0; i < tswitch->getCaseCount(); i++)
+                blocks.add(tswitch->getCaseBlock(i));
+
+            IRInst* tmp = inst1->getParent();
```
Comment (csyonghe) [inline source/slang/slang-ir-reachability.cpp:81]: You may want to call getBlock instead of getParent here. GetBlock returns the enclosing block of the instruction. For instructions in spirv asm blocks, their parent is the spirv inst instead of the enclosing block.

File: source/slang/slang-ir-reachability.cpp (hunk 1 above)
Comment (csyonghe) [reply source/slang/slang-ir-reachability.cpp:88]: In that case we should be extending isBlockReachable to properly handle target switch insts along with other control flow.

File: source/slang/slang-ir-reachability.cpp [hunk 3]
```diff
@@ -54,6 +55,45 @@ namespace Slang
         }
 
     }
+    
+    bool ReachabilityContext::isInstReachable(IRInst* inst1, IRInst* inst2)
+    {
+        // If inst1 and inst2 are in the same block,
+        // we test if inst2 appears after inst1.
```
Comment (jkwak-work) [inline source/slang/slang-ir-reachability.cpp:62]: If a certain order is expected for the input parameters, the naming may be better with `from` and `to`.
```
bool ReachabilityContext::isInstReachable(IRInst* from, IRInst* to)
```
Or `before` and `after`; or `top` and `down`... I don't know.

File: source/slang/slang-ir-reachability.cpp [hunk 4]
```diff
@@ -54,6 +55,45 @@ namespace Slang
...
+        if (inst1->getParent() == inst2->getParent())
```
Comment (jkwak-work) [inline source/slang/slang-ir-reachability.cpp:63]: Shouldn't this call `getBlock`?
That's what the comment above says.
```
if (inst1->getBlock() == inst2->getBlock())
```

File: source/slang/slang-ir-use-uninitialized-values.cpp [hunk 5]
```diff
@@ -37,30 +37,30 @@ namespace Slang
             || (inst->m_op == kIROp_Var);
     }
 
-    static bool isUndefinedParam(IRParam* param)
+    static bool isPotentiallyUnintended(IRParam* param)
```
Comment (jkwak-work) [inline source/slang/slang-ir-use-uninitialized-values.cpp:40]: I think this name much better.
We often use a naming like `maybeXXX`, and I like `Potentially` too.
But shouldn't it be `isPotentiallyUninitialized`?
I am not sure what Unintended means in this context.

File: tests/diagnostics/uninitialized-out-parameters.slang [hunk 6]
```diff
@@ -47,5 +47,48 @@ void control_flow_defined(bool b, out float y)
...
+    case spirv:
+        {
+            if ((x & 2) == 0)
+                v = half(1);
+            else
+                v = half(1);
+            return;
+        }
+    }
+}
```
Comment (jkwak-work) [inline tests/diagnostics/uninitialized-out-parameters.slang:91]: I am wondering about a case like following,
```
void target_switching_blocks(uint x, out half v)
{
    __target_switch
    {
    case hlsl:
        return; // <=== WOULD IT PRINT THE WARNING WHEN TARGETTING HLSL?
    case glsl:
    case spirv:
        {
            v = half(1);
            return;
        }
    }
}
```

File: tests/diagnostics/uninitialized-out-parameters.slang [hunk 7]
```diff
@@ -47,5 +47,48 @@ void control_flow_defined(bool b, out float y)
     return;
 }
 
+// Specialized handling for intrinsic asm
+void instrincs_defined(float x, out float exp)
+{
+    __intrinsic_asm "frexp";
+}
```
Comment (jkwak-work) [inline tests/diagnostics/uninitialized-out-parameters.slang:54]: FYI, the following case should print a warning but I assume that it will be not easy to implement.
```
void instrincs_defined(float x, out float exp)
{
    __intrinsic_asm "frexp(x)";
}
```
We will need to ignore this case for now.

File: tests/diagnostics/uninitialized-out-parameters.slang [hunk 8]
```diff
@@ -47,5 +47,48 @@ void control_flow_defined(bool b, out float y)
...
+
+// Specialized handling for target switches
+void target_switching_undefined(float x, out float exp)
+{
+    //CHK-DAG: warning 41018: returning without initializing out parameter 'exp'
+    __target_switch {}
```
Comment (jkwak-work) [inline tests/diagnostics/uninitialized-out-parameters.slang:60]: Using `-DAG` looks dangerous, because it may match to a warning from totally different part of this test.
You can include the line number into the pattern.

File: source/slang/slang-ir-reachability.cpp (hunk 2 above)
Comment (venkataram-nv) [reply source/slang/slang-ir-reachability.cpp]: Where is the `getBlock()` method? It doesn't look like `IRInst` has that.

File: source/slang/slang-ir-reachability.cpp (hunk 1 above)
Comment (venkataram-nv) [reply source/slang/slang-ir-reachability.cpp]: I have simplified the logic for this part to rely on `isBlockReachable` more. We can't exactly extended `isBlockReachable` to handle switches and control flow because then it would relax the reachability to the point where something within the control flow's block is accessible after the control flow.

File: tests/diagnostics/uninitialized-out-parameters.slang (hunk 6 above)
Comment (venkataram-nv) [reply tests/diagnostics/uninitialized-out-parameters.slang:91]: I does print the warning, although not exclusively for hlsl targets. The warning here would be returning without initializing `v`

Should we only warn for the particular target? The diagnostics are being evaluated at the IR level so I'm not sure.

File: source/slang/slang-ir-use-uninitialized-values.cpp (hunk 5 above)
Comment (venkataram-nv) [reply source/slang/slang-ir-use-uninitialized-values.cpp:40]: The refactor is kind of alluding the inclusion of checking `inout` parameters. In that case, if the user never writes to the parameter, they might not have meant to keep the parameter `inout` -- so making the parameter `inout` is unintended.

File: source/slang/slang-ir-reachability.cpp [hunk 9]
```diff
@@ -52,17 +53,51 @@ namespace Slang
...
+            IRBlock* upper = as<IRBlock>(from->getParent());
+            for (Slang::UInt i = 0; i < tswitch->getCaseCount(); i++)
+            {
+                IRBlock* caseBlock = tswitch->getCaseBlock(i);
+                if (caseBlock == upper || isBlockReachable(caseBlock, upper))
+                    return true;
+            }
+        }
+
+        return isBlockReachable(as<IRBlock>(from->getParent()), as<IRBlock>(to->getParent()));
```
Comment (csyonghe) [inline source/slang/slang-ir-reachability.cpp:85]: `getBlock(from)` and `getBlock(to)`.

File: source/slang/slang-ir-reachability.cpp [hunk 10]
```diff
@@ -52,17 +53,51 @@ namespace Slang
...
+            {
+                if (inst == to)
+                    return true;
+            }
+        }
 
+        // Special cases
+        
+        // Target switches; treat as reachable from any of its cases
+        if (auto tswitch = as<IRTargetSwitch>(to))
```
Comment (csyonghe) [inline source/slang/slang-ir-reachability.cpp:74]: Can we move this to `isBlockReachable`?

Comment (venkataram-nv) [reply source/slang/slang-ir-reachability.cpp:75]: I don't think that works for this case since `tswitch` is a specific instruction. We need to check against the case blocks of `tswitch` rather than `getBlock(tswitch)`

File: tests/diagnostics/uninitialized-out-parameters.slang (hunk 6 above)
Comment (jkwak-work) [reply tests/diagnostics/uninitialized-out-parameters.slang:91]: Aha. I guess, we cannot print the warning if the checking is happening at ir level.

File: tests/diagnostics/uninitialized-out-parameters.slang [hunk 11]
```diff
@@ -3,32 +3,32 @@
 // Using before assigning
 float regular_undefined_use(out float3 v)
 {
-    //CHK-DAG: warning 41015: use of uninitialized out parameter 'v'
+    //CHK-DAG: (7): warning 41015: use of uninitialized out parameter 'v'
```
Comment (jkwak-work) [inline tests/diagnostics/uninitialized-out-parameters.slang:6]: I am AFK now but there is a better way to check the line numbers. Something like line+1 or something.
Please check the other tests for an example.

File: source/slang/slang-ir-reachability.cpp (hunk 10 above)
Comment (venkataram-nv) [reply source/slang/slang-ir-reachability.cpp]: So I have moved this special case away from `Reachability` and to the `slang-ir-use-uninitialized-values.cpp` file. There is no problem with target switches with respect to reachability, but it adds a case that needs handling for the warnings. For example we want to warn cases like:
```
void foo(float x, out float exp)
{
      __target_switch {}
}
```
But not cases like this:
```
void foo(float x, out float exp)
{
     __target_switch
     {
     case glsl: __intrinsic_asm("frexp");
     }
}
```
To handle the first case, we treat target switches as loads (same thing is done for a return) so they are checked at the end. If it is a terminal instruction (as with both cases above), then in order for the target switch to be OK there needs to be some instruction in (any of) its case blocks which stores into the out parameter. This is why I wanted to establish that the target switch should be reachable from its nested instructions.

In the ordinary use case, thought, I don't think this is appropriate, which is why I have moved that code back into `slang-ir-use-uninitialized-values.cpp`.

Comment (csyonghe) [reply source/slang/slang-ir-reachability.cpp]: I think the problem is not about target switch, it is just that we need to turn off the warning checking for functions parameters if the function body has __intrinsic_asm.

Comment (venkataram-nv) [reply source/slang/slang-ir-reachability.cpp]: But there should still be some check for control flows path even there is a __intrinsic_asm. For example if users early exit before the __intrinsic_asm instruction, there should be some warning. Target switches are involved in this because they can appear as block terminal instructions, and this was previously not considered.

Comment (venkataram-nv) [reply source/slang/slang-ir-reachability.cpp]: With some changes, the TLDR for this PR is:
* Add consideration for `target_switch` as block terminator instructions
* Manually sift through `__intrinsic_asm` instructions found in all blocks
And the rest of the main logic (`cancelLoadStores`) remains the same.

File: source/slang/slang-ir-use-uninitialized-values.cpp [hunk 12]
```diff
@@ -277,30 +285,47 @@ namespace Slang
...
 
-    static void cancelLoads(ReachabilityContext &reachability, const List<IRInst*>& stores, List<IRInst*>& loads)
+    static bool canStoreReachLoad(ReachabilityContext& reachability, IRInst* store, IRInst* load)
+    {
+        if (reachability.isInstReachable(store, load))
+            return true;
+
+        // Special cases
+        
+        // Target switches; treat as reachable from any of its cases
```
Comment (ArielG-NV) [inline source/slang/slang-ir-use-uninitialized-values.cpp:295]: > Target switches are by default not reachable by each of their cases

Would it be possible to add an option/Flag to `ReachabilityContext` such that we don't need to have code that handles an "exception" to the rule of reachability explicitly for `__target_switch`?

Comment (csyonghe) [reply source/slang/slang-ir-use-uninitialized-values.cpp:295]: What if `load` is an inst inside a target switch case? It doesn't seem like this logic will capture that.

Comment (venkataram-nv) [reply source/slang/slang-ir-use-uninitialized-values.cpp:295]: > Would it be possible to add an option/Flag to ReachabilityContext
Yes, we could, but I'm not sure its absolutely necessary for the core purpose of `ReachabilityContext`. The exception with target switch for reachability is only for the warnings so far.

> What if load is an inst inside a target switch case? It doesn't seem like this logic will capture that.
If there was an instruction explicitly loading from the parameter then it would be found in the `IRUse` chain for it, so it doesn't need to be handled in a special way.

Comment (csyonghe) [reply source/slang/slang-ir-use-uninitialized-values.cpp:295]: From what I read the code is doing here: if `load` is a target switch and we want to check if a `store` can reach this target switch, then we go through all the target switch cases and check if that store is reachable from the case block, if it is reachable from any case block, then we say yes?

I don't quite understand  the logic behind this. We are checking if a store can reach a target switch, but then why do we do so by checking if the case blocks can reach the store? It seems reversed...

Comment (venkataram-nv) [reply source/slang/slang-ir-use-uninitialized-values.cpp:295]: It is indeed reversed, and that is because in this case the store instructions would appear after the target switch (and its case blocks). If we did `isBlockReachable(upper, caseBlock)` instead, it would always be false in this case because a case block is not reachable from an inner instruction.

Comment (csyonghe) [reply source/slang/slang-ir-use-uninitialized-values.cpp:295]: Then why `canStoreReachLoad` is called with `load = target_switch`? If `load` is before `store`, then the function should always return false as the name suggests.

Comment (venkataram-nv) [reply source/slang/slang-ir-use-uninitialized-values.cpp:295]: We want to make an exception for `__target_switch`s as we want to include a warning if they are the last instruction of block (and if they don't initialize the out parameter). Just like for `IRReturn`s we treat `__target_switch` as a load instruction so that it goes through `cancelLoadStores`. For it to work out in `cancelLoadStores` we do special handling in `canStoreReachLoad` where for target switches only we allow the store to appear after the load.

I could refactor the method name to something like `canStoreCancelLoad` if it makes more sense.

File: source/slang/slang-ir-reachability.h [hunk 13]
```diff
@@ -16,23 +16,8 @@ struct ReachabilityContext
     ReachabilityContext() = default;
     ReachabilityContext(IRGlobalValueWithCode* code);
 
+    bool isInstReachable(IRInst* from, IRInst* to);
```
Comment (csyonghe) [inline source/slang/slang-ir-reachability.h:19]: We can revert this change and move `isInstReachable` to its original place?

Comment (venkataram-nv) [reply source/slang/slang-ir-reachability.h:19]: We can but after switching to `getBlock` calls we would also need to `#include "slang-ir-util.h"` in `slang-ir-reachabilty.h` is that fine?

Comment (csyonghe) [reply source/slang/slang-ir-reachability.h:19]: Ah, I see, moving it here is fine.

**Review by csyonghe — APPROVED** [review]
